### PR TITLE
Add for_each_concurrent

### PR DESCRIPTION
**Pull request by cramertj**

### CODE REVIEW

File: futures-util/src/stream/mod.rs [hunk 1]
```diff
@@ -599,6 +601,30 @@ pub trait StreamExt: Stream {
...
 
+    /// Runs this stream to completion, executing the provided closure for each
+    /// element on the stream. This is similar to `for_each` but may begin
+    /// processing an element while previous elements are still being processed.
+    ///
+    /// When this stream successfully resolves to an item, the closure will be
+    /// called to produce a future. That future will then be added to
+    /// the set of futures to resolve.
+    ///
+    /// The returned value is a `Future` where the `Item` type is `()` and
```
Comment (aturon) [inline futures-util/src/stream/mod.rs]: The docs don't seem to match, since you're actually returning back the stream.

**Comment by aturon** [comment]
r+ modulo minor nit

File: futures-util/src/stream/mod.rs (hunk 1 above)
Comment (cramertj) [reply futures-util/src/stream/mod.rs]: oh good catch! that's wrong on for_each, too

Comment (cramertj) [reply futures-util/src/stream/mod.rs]: Fixed.

File: futures-util/src/stream/for_each_concurrent.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+use futures_core::{Async, Future, IntoFuture, Poll, Stream};
+use futures_core::task;
+
+use super::futures_unordered::FuturesUnordered;
+
+/// A stream combinator which executes a unit closure over each item on a
+/// stream concurrently.
+///
+/// This structure is returned by the `Stream::for_each` method.
```
Comment (tikue) [inline futures-util/src/stream/for_each_concurrent.rs]: `Stream::for_each` -> `Stream::for_each_concurrent`

**Comment by tikue** [comment]
When would one prefer this over `stream.for_each(|item| tokio::spawn(f(item)))`?

**Pull request merged**